--- xboard-4.2.6/xboard.texinfo.in~	Mon Nov 11 12:51:09 2002
+++ xboard-4.2.6/xboard.texinfo.in	Mon Nov 11 12:51:09 2002
@@ -15,10 +15,10 @@
 
 @ifinfo
 @format
-INFO-DIR-SECTION Games
-START-INFO-DIR-ENTRY
-* xboard: (xboard).         An X Window System graphical chessboard.
-END-INFO-DIR-ENTRY
+@dircategory Games:
+@direntry
+* xboard: (xboard).			An X11 graphical chessboard
+@end direntry
 @end format
 @end ifinfo
 
