--- xboard-4.2.7/xboard.c.xtname	2005-06-03 07:49:06.000000000 +0800
+++ xboard-4.2.7/xboard.c	2005-06-03 07:49:45.000000000 +0800
@@ -4790,9 +4790,7 @@ void EditCommentCallback(w, client_data,
     int j;
     Widget edit;
 
-    j = 0;
-    XtSetArg(args[j], XtNlabel, &name);  j++;
-    XtGetValues(w, args, j);
+    name = XtName(w);
 
     if (strcmp(name, "ok") == 0) {
 	edit = XtNameToWidget(editShell, "*form.text");
--- xboard-4.2.7/xedittags.c.xtname	2003-10-28 03:21:00.000000000 +0800
+++ xboard-4.2.7/xedittags.c	2005-06-03 07:49:45.000000000 +0800
@@ -105,13 +105,7 @@ void TagsCallback(w, client_data, call_d
      Widget w;
      XtPointer client_data, call_data;
 {
-    String name;
-    Arg args[16];
-    int j;
-
-    j = 0;
-    XtSetArg(args[j], XtNlabel, &name);  j++;
-    XtGetValues(w, args, j);
+    String name = XtName(w);
 
     if (strcmp(name, "close") == 0) {
 	TagsPopDown();
@@ -131,9 +125,7 @@ void EditTagsCallback(w, client_data, ca
     int j;
     Widget textw;
 
-    j = 0;
-    XtSetArg(args[j], XtNlabel, &name);  j++;
-    XtGetValues(w, args, j);
+    name = XtName(w);
 
     if (strcmp(name, "ok") == 0) {
 	textw = XtNameToWidget(editTagsShell, "*form.text");
